subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,716 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(net_coap, CONFIG_COAP_LOG_LEVEL);
+
+#include <zephyr/net/socket.h>
+
+#include <zephyr/net/coap.h>
+#include <zephyr/net/coap_client.h>
+
+#define COAP_VERSION 1
+#define COAP_PATH_ELEM_DELIM '/'
+#define COAP_PATH_ELEM_QUERY '?'
+#define COAP_PATH_ELEM_AMP '&'
+#define COAP_SEPARATE_TIMEOUT 6000
+#define DEFAULT_RETRY_AMOUNT 5
+#define BLOCK1_OPTION_SIZE 4
+#define PAYLOAD_MARKER_SIZE 1
+
+static int coap_client_schedule_poll(struct coap_client *client, int sock,
+				     struct coap_client_request *req)
+{
+	client->fd = sock;
+	client->coap_request = req;
+
+	k_sem_give(&client->coap_client_recv_sem);
+	atomic_set(&client->coap_client_recv_active, 1);
+
+	return 0;
+}
+
+static int send_request(int sock, const void *buf, size_t len, int flags,
+			const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	if (addrlen == 0) {
+		return sendto(sock, buf, len, flags, NULL, 0);
```

Comment:
Note that `sendto` returns a positive value indicating the number of bytes sent.  This means that whatever calls `send_request` needs to know that.